### DISCUSSION TITLE
Is sympy dsolve still a little broken?

### DISCUSSION THREAD

**Discussion by DukeDoc99** [sympy/sympy#24841] [Q&A]
I have been trying to use sympy for things that I previously used other unnamed symbolic solvers, and I keep hitting spaces where the answers are just...wrong.  Current case is setting up and solving a system of mixed differential and linear algebra equations for a circuit.  dsolve solves some of the variables correctly but some do not satisfy the actual equations given to dsolve and there seems to be a part of the solution (specifically an e^{-t} part) that is coming from thin air?  Full example at:

https://colab.research.google.com/drive/1QHtDJOxBNN8cs9AeUkLRL9EBhA66lTVO?usp=sharing 

Also - sympy is fine with these specific numbers, but if you change the numerical value of R2 to 4, sympy chokes (which I think is a continuing issue with underdamped differential equations).  Changing R2 to something larger than 6, like 8, will get an overdamped answer, but again the erroneous e^{-t} appears in parts of the solution and the equations are not all satisfied...

**Comment by oscarbenjamin** [comment]
It's fine to post a link to something like a notebook but please actually include the code directly. Don't presume to expect other people to spend the time putting the code together if you are not gong to bother yourself. (This applies to any kind of forum whether it's github, a mailing list, stackoverflow etc.)

Also a title like `Is sympy dsolve still a little broken?` is hyperbolic. There are lots of cases where `dsolve` works well but yes it has bugs. Your example is a very particular kind of system of equations which `dsolve` is not designed for: DAEs. Perhaps it should reject those (I think it used to) or perhaps it is straight-forward to adapt the solver to handle simple cases like this where the auxiliary equations are linear.

Here is your code:
```python
import sympy as sym
import numpy as np
import matplotlib.pyplot as plt
PAR = lambda R1, R2: R1 * R2 / (R1 + R2)

R1, R2, R3, C, L, vadc, vbdc, t = sym.var('R1, R2, R3, C, L, v_{a\,dc}, v_{b\,dc},t')
va = sym.Function("v_a")
vb = sym.Function("v_b")
iL = sym.Function("i_L")
vC = sym.Function("v_C")
vx = sym.Function("v_x")
vy = sym.Function("v_y")
unks = [iL(t), vC(t), vx(t), vy(t)]
knowns =     R1, R2, R3,   C,   L, vadc, vbdc, va(t), vb(t)
knownvals = [ 6,  6,  6, 1/8, 1/2,   45,   15,    45,    15]

iL0 = iL(t).subs(t, 0)
vC0 = vC(t).subs(t, 0)
vx0 = vx(t).subs(t, 0)
vy0 = vy(t).subs(t, 0)
unksSS = [iL0, vC0, vx0, vy0]
eqn1SS = (vC0 - vadc) / R1 + (vC0 - vbdc) / R2
eqn2SS = (vbdc - vC0) / R2 + (vbdc - vadc) / R3 + iL0
eqn3SS = (vx0 - vC0) / R1 + (vx0 - vy0) / R3
eqn4SS = (vy0 - vx0) / R3 + (vy0 - vC0) / R2 + iL0
eqnsSS = [eqn1SS, eqn2SS, eqn3SS, eqn4SS]
solnSS = sym.solve(eqnsSS, unksSS)

def makesubs(solution, vartosub, valtosub):
    subsol = solution.copy()
    sublist = list(zip(vartosub, valtosub))
    if type(subsol)==list:
        subsol = makedict(subsol)
    for var in subsol:
        subsol[var] = subsol[var].subs(sublist)
    return subsol

def makesubseqnlist(equations, vartosub, valtosub):
    sublist = list(zip(vartosub, valtosub))
    esub = []
    for eqn in equations:
      esub.append(eqn.subs(sublist))
    return esub

solnSSnum = makesubs(solnSS, knowns, knownvals)

# Define equations
eq1 = sym.Eq((vC(t)-vx(t))/R1 + (vC(t)-vy(t))/R2 + C * sym.Derivative(vC(t), t), 0)
eq2 = sym.Eq(vy(t) - vb(t), L * sym.Derivative(iL(t), t))
eq3 = sym.Eq((vx(t)-vC(t))/R1 + (vx(t) - vy(t)) / R3, 0)
eq4 = sym.Eq((vy(t) - vx(t))/R3 + (vy(t)-vC(t))/R2 + iL(t), 0)

# Collect equations
eqns = [eq1, eq2, eq3, eq4]

neqns = makesubseqnlist(eqns, knowns, knownvals)
neqns

eqs = neqns
funcs = unks
ics = solnSSnum

# Solve and show raw solution (will be a list)
solnlist = sym.dsolve(eqs, funcs, ics=ics)
```
Your equations are:
```python
In [2]: for eq in eqs: pprint(eq)
v_C(t)   vₓ(t)   v_y(t)         d             
────── - ───── - ────── + 0.125⋅──(v_C(t)) = 0
  3        6       6            dt            
                  d         
v_y(t) - 15 = 0.5⋅──(i_L(t))
                  dt        
  v_C(t)   vₓ(t)   v_y(t)    
- ────── + ───── - ────── = 0
    6        3       6       
         v_C(t)   vₓ(t)   v_y(t)    
i_L(t) - ────── - ───── + ────── = 0
           6        6       3 
```
The second two equations here are not ODEs so this is actually a system of DAEs (differential algebraic equations) which `dsolve` is not really intended to support so I'm not sure exactly how `dsolve`'s system solver would attempt something like this. I confirm though that the answer given by `dsolve` is incorrect though.

There are several ways to approach DAEs. One way is to differentiate the algebraic equations:
```python
In [4]: eqs2 = eqs[:2] + [eq.lhs.diff(t) for eq in eqs[2:]]

In [5]: for eq in eqs2: pprint(eq)
v_C(t)   vₓ(t)   v_y(t)         d             
────── - ───── - ────── + 0.125⋅──(v_C(t)) = 0
  3        6       6            dt            
                  d         
v_y(t) - 15 = 0.5⋅──(i_L(t))
                  dt        
  d            d           d         
  ──(v_C(t))   ──(vₓ(t))   ──(v_y(t))
  dt           dt          dt        
- ────────── + ───────── - ──────────
      6            3           6     
             d            d           d         
             ──(v_C(t))   ──(vₓ(t))   ──(v_y(t))
d            dt           dt          dt        
──(i_L(t)) - ────────── - ───────── + ──────────
dt               6            6           3    
```
Now `dsolve` gives the correct answer:
```python
In [6]: sol2 = dsolve(eqs2, funcs, ics=ics)

In [7]: sol2
Out[7]: 
⎡              -4.0⋅t                        -4.0⋅t                                      -4.0⋅t⎤
⎣i_L(t) = 7.5⋅ℯ      , v_C(t) = 15.0 + 15.0⋅ℯ      , vₓ(t) = 15.0, v_y(t) = 15.0 - 15.0⋅ℯ      ⎦

In [8]: sol2subs = {e.lhs:e.rhs for e in sol2}

In [9]: eqs[0].subs(sol2subs)
Out[9]: 
      d ⎛             -4.0⋅t⎞        -4.0⋅t    
0.125⋅──⎝15.0 + 15.0⋅ℯ      ⎠ + 7.5⋅ℯ       = 0
      dt                                       

In [10]: eqs[0].subs(sol2subs).doit()
Out[10]: True

In [11]: eqs[1].subs(sol2subs).doit()
Out[11]: True

In [12]: eqs[2].subs(sol2subs).doit()
Out[12]: True

In [13]: eqs[3].subs(sol2subs).doit()
Out[13]: True

In [14]: checkodesol(eqs2, sol2)
Out[14]: (True, [0, 0, 0, 0])

In [15]: ics
Out[15]: {i_L(0): 15/2, v_C(0): 30, vₓ(0): 15, v_y(0): 0}

In [16]: [s.subs(t, 0) for s in sol2]
Out[16]: [i_L(0) = 7.5, v_C(0) = 30.0, vₓ(0) = 15.0, v_y(0) = 0]
```
Another approach would be to use an algebraic solver like `linsolve` to eliminate `vx` and `vy`:
```python
In [19]: solve(eqs, [iL(t).diff(t), vC(t).diff(t), vx(t), vy(t)])
Out[19]: 
⎧                                                           d                                      
⎨vₓ(t): -2.0⋅i_L(t) + v_C(t), v_y(t): -4.0⋅i_L(t) + v_C(t), ──(i_L(t)): -8.0⋅i_L(t) + 2.0⋅v_C(t) - 
⎩                                                           dt                                     

      d                      ⎫
30.0, ──(v_C(t)): -8.0⋅i_L(t)⎬
      dt                     ⎭

In [20]: asol = solve(eqs, [iL(t).diff(t), vC(t).diff(t), vx(t), vy(t)])

In [23]: eqs3 = [Eq(f, asol[f]) for f in [iL(t).diff(t), vC(t).diff(t)]]

In [24]: eqs3
Out[24]: 
⎡d                                             d                       ⎤
⎢──(i_L(t)) = -8.0⋅i_L(t) + 2.0⋅v_C(t) - 30.0, ──(v_C(t)) = -8.0⋅i_L(t)⎥
⎣dt                                            dt                      ⎦

In [26]: ics
Out[26]: {i_L(0): 15/2, v_C(0): 30, vₓ(0): 15, v_y(0): 0}

In [27]: ics3 = {vC(0):ics[vC(0)], iL(0):ics[iL(0)]}

In [28]: dsolve(eqs3, ics=ics3)
Out[28]: 
⎡              -4.0⋅t                        -4.0⋅t⎤
⎣i_L(t) = 7.5⋅ℯ      , v_C(t) = 15.0 + 15.0⋅ℯ      ⎦
```
The other two variables are then given in terms of these from the algebraic solution:
```python
In [29]: asol[vx(t)]
Out[29]: -2.0⋅i_L(t) + v_C(t)

In [30]: asol[vy(t)]
Out[30]: -4.0⋅i_L(t) + v_C(t)
```

**Reply by oscarbenjamin** [reply]
Note that DAEs are in general much harder to solve than ODEs:
https://en.wikipedia.org/wiki/Differential-algebraic_system_of_equations
http://www.scholarpedia.org/article/Differential-algebraic_equations

A case like this is easier because it is at least simple to solve for the variables that do have derivatives in the equations meaning that this can be reduced to solving a system of ODEs. However in your example you give initial conditions for all 4 variables but transforming to a pair of ODEs means that the general solution will have only two degrees of freedom so there is no longer any guarantee of the existence of a solution to the initial value problem (the initial conditions might contradict the algebraic equations).

**Reply by oscarbenjamin** [reply]
> Don't presume to expect other people to spend the time putting the code together if you are not gong to bother yourself.

Sorry if this comes across too strong. I should have worded that more delicately. I realise that you did put the time and effort to make a good notebook to illustrate the point so thank you for doing that.

I just spend a lot of time looking at things like this and I get frustrated if it takes more time than necessary for me to extract the code. It should be fairly obvious that the first thing anyone would want to do when investigating something like this is to run the code themselves so it should generally be given in a form that can be easily copy/pasted.

Ideally examples should also be simplified with extraneous details removed like this. So the ideal thing would be like this (or perhaps even with the variables renamed):
```python
from sympy import *

v_x, v_y, v_C, i_L = symbols('v_x, v_y, v_C, i_L', cls=Function)
t = symbols('t')

eqs = [
    Eq(v_C(t)/3 - v_x(t)/6 - v_y(t)/6 + 0.125*Derivative(v_C(t), t), 0),
    Eq(v_y(t) - 15, 0.5*Derivative(i_L(t), t)),
    Eq(-v_C(t)/6 + v_x(t)/3 - v_y(t)/6, 0),
    Eq(i_L(t) - v_C(t)/6 - v_x(t)/6 + v_y(t)/3, 0),
]

funcs = [v_x(t), v_y(t), v_C(t), i_L(t)]
ics = {i_L(0): 15/2, v_C(0): 30, v_x(0): 15, v_y(0): 0}

# Correct solution:
expected = [
    Eq(v_x(t), 15.0),
    Eq(v_y(t), 15.0 - 15.0*exp(-4.0*t)),
    Eq(v_C(t), 15.0 + 15.0*exp(-4.0*t)),
    Eq(i_L(t), 7.5*exp(-4.0*t)),
]

# Incorrect: returns ...
sol = dsolve(eqs, funcs, ics=ics)
print(sol)
```

Also it is better not to rely on linking to external things like notebooks because they might be removed later. The code in question needs to be right here so it will be preserved in future. Lots of old SymPy issues now have stale links making them not reproducible any more.

**Comment by oscarbenjamin** [comment]
It looks like `_classify_linear_system` misclassifies this as being a "type 2" system meaning constant coefficients non-homogeneous of the form `X' + A*X = b`. Actually it is `A0*X' + A*X = b` where `A0` is:
```python
(Pdb) As[0]
Matrix([
[1, 0, 0, 0],
[0, 1, 0, 0],
[0, 0, 0, 0],
[0, 0, 0, 0]])
```
It seems that `_classify_linear_system` assumes that `As[0]` will always be the identity matrix because the equations were already solved for the highest derivative of each independent variable. Just in this case that "highest derivative" is not necessarily a derivative because some variables don't have derivatives.

Fixing that it seems that the weak component solver can correctly separate this into a pair of ODEs and two dependent (already solved) algebraic equations. It just needs to be taught to understand that some non-differential equations might be there as well. This diff makes it so that `dsolve` can handle this case:
```diff
diff --git a/sympy/solvers/ode/systems.py b/sympy/solvers/ode/systems.py
index c0cbd51..0e02f30 100644
--- a/sympy/solvers/ode/systems.py
+++ b/sympy/solvers/ode/systems.py
@@ -1413,7 +1413,26 @@ def _classify_linear_system(eqs, funcs, t, is_canon=False):
     order = _get_func_order(eqs, funcs)
     system_order = max(order[func] for func in funcs)
     is_higher_order = system_order > 1
-    is_second_order = system_order == 2 and all(order[func] == 2 for func in funcs)
+    is_second_order = all(order[func] == 2 for func in funcs)
+
+    if any(order[func] == 0 for func in funcs):
+        # The system contains some purely algebraic equations. The general ODE
+        # systems solvers do not incorporate such equations but the weak component
+        # solver should be able to separate them from the differential equations.
+        #
+        # This happens if we have e.g.:
+        #
+        # (1)  x' = x + y + z
+        # (2)  y' = x - y
+        # (3)  z  = t + x + y
+        #
+        # The weak component solver will divide this into two components (1, 2)
+        # and (3). The first system (1, 2) is all first order. Once solved its
+        # solutions can be substituted directly into (3) which immediately
+        # gives z. The strong component solver will accept (3) as being already
+        # the solution for z. We return None now so that weak component solver
+        # can try again and return components in the form expected here.
+        return None
 
     # Not adding the check if the len(func.args) for
     # every func in funcs is 1
@@ -1696,6 +1715,25 @@ def _higher_order_ode_solver(match):
 def _strong_component_solver(eqs, funcs, t):
     from sympy.solvers.ode.ode import dsolve, constant_renumber
 
+    # In the case of purely algebraic equations we might actually just have
+    # something like:
+    #
+    #  x = something
+    #  y = something
+    #
+    # If that is the case then we can just return the equation as is since it
+    # already gives the solution. Algebraic equations in the orginal input will
+    # be in this form already because an earlier step solves for them
+    # algebraically. The weak component solver will separate such equations
+    # from the differential equations so it should not be necessary to handle
+    # both algebraic and differential equations in the same component.
+
+    def is_solved(eq):
+        return eq.lhs in funcs and not eq.rhs.has_xfree(set(funcs))
+
+    if all(is_solved(eq) for eq in eqs):
+        return eqs
+
     match = _classify_linear_system(eqs, funcs, t, is_canon=True)
     sol = None
 
@@ -1726,7 +1764,14 @@ def _strong_component_solver(eqs, funcs, t):
 
 
 def _get_funcs_from_canon(eqs):
-    return [eq.lhs.args[0] for eq in eqs]
+    #
+    # We need to check if the lhs is a derivative in case there are equations
+    # without derivatives e.g. like:
+    #
+    #  x' = x + y
+    #  y = x + 1
+    #
+    return [eq.lhs.args[0] if eq.lhs.is_Derivative else eq.lhs for eq in eqs]
 
 
 # Returns: List of Equations(a solution)
@@ -2094,7 +2139,8 @@ def dsolve_system(eqs, funcs=None, t=None, ics=None, doit=False, simplify=True):
         '''))
 
     if t is None:
-        t = list(list(eqs[0].atoms(Derivative))[0].atoms(Symbol))[0]
+        assert len(funcs[0].args) == 1
+        t = funcs[0].args[0]
 
     sols = []
     canon_eqs = canonical_odes(eqs, funcs, t)
```
With those changes we have:
```python
In [3]: dsolve(eqs, funcs, ics=ics)
Out[3]: 
⎡              -4.0⋅t                        -4.0⋅t                                      -4.0⋅t⎤
⎣i_L(t) = 7.5⋅ℯ      , v_C(t) = 15.0 + 15.0⋅ℯ      , vₓ(t) = 15.0, v_y(t) = 15.0 - 15.0⋅ℯ      ⎦
```

**Reply by oscarbenjamin** [reply]
I've opened gh-24848 with these changes.

**Reply by DukeDoc99** [reply]
Thank you for all of the guidance - on how to recast the problem correctly, on how to properly and more effectively ask (and summarize) questions, and on how to potentially fix the original issue!

**Reply by oscarbenjamin** [reply]
Thanks for reporting this. It looks like the general mechanism in `dsolve_system` can effectively support solving simple DAE systems like this. It was designed in a sufficiently general way that solvable algebraic systems can almost be handled naturally just as ODEs of "order zero". Just it wasn't written or tested with those cases in mind so needs tweaks in a few places to make it work properly because there is code that assumes that all equations are ODEs.
